hq_superset/models.py

```diff
@@ -0,0 +1,34 @@
+from dataclasses import dataclass
+from typing import Any
+
+from .utils import cast_data_for_table, get_hq_database
+
+
+@dataclass
+class DataSetChange:
+    data_source_id: str
+    doc_id: str
+    data: list[dict[str, Any]]
+
+    def update_dataset(self):
+        database = get_hq_database()
+        try:
+            sqla_table = next((
+                table for table in database.tables
+                if table.table_name == self.data_source_id
+            ))
+        except StopIteration:
+            raise ValueError(f'{self.data_source_id} table not found.')
+        table = sqla_table.get_sqla_table_object()
+
+        with (
+            database.get_sqla_engine_with_context() as engine,
+            engine.connect() as connection,
+            connection.begin()  # Commit on leaving context
+        ):
+            delete_stmt = table.delete().where(table.c.doc_id == self.doc_id)
+            connection.execute(delete_stmt)
+            if self.data:
+                rows = list(cast_data_for_table(self.data, table))
+                insert_stmt = table.insert().values(rows)
+                connection.execute(insert_stmt)
```

Comment:
is this deleting and then inserting the row for the same case instead of an update?
Not that I am opposed to the approach, just trying to understand the reasoning behind it.

Reply:
I wrangled with this for a long time. SQLAlchemy's [Postgres dialect supports upsert](https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#insert-on-conflict-upsert). But there is a catch: The "doc_id" column must be indexed in order to use it as a constraint in the upsert's "where" clause.

That table is created by Pandas, and Pandas can't create an index on that column. You can create the table first, and then dump Pandas data into it, but I couldn't get Pandas to create the rest of columns and populate the table correctly. Maybe I was missing something. Eventually I decided to just delete and then insert.

Next time I have a few more hours to spend on it, I'll go back to it. Or we can make a follow-up ticket for nice-to-haves.

Reply:
thanks @kaapstorm 
the current approach sounds good to me